src/Common/Threading.cpp

```diff
@@ -0,0 +1,486 @@
+    {
+        bool res = CancelToken::local().wait(lowerAddress(&address), lowerValue(value));
+        value = address.load();
+        return res;
+    }
+
+    inline void wakeLowerOne(std::atomic<UInt64> & address)
+    {
+        syscall(SYS_futex, lowerAddress(&address), FUTEX_WAKE_PRIVATE, 1, nullptr, nullptr, 0);
+    }
+
+    // inline void wakeLowerAll(std::atomic<UInt64> & address)
+    // {
+    //     syscall(SYS_futex, lowerAddress(&address), FUTEX_WAKE_PRIVATE, INT_MAX, nullptr, nullptr, 0);
+    // }
+
+    inline void waitUpperFetch(std::atomic<UInt64> & address, UInt64 & value)
+    {
+        futexWait(upperAddress(&address), upperValue(value));
+        value = address.load();
+    }
+
+    inline bool cancellableWaitUpperFetch(std::atomic<UInt64> & address, UInt64 & value)
+    {
+        bool res = CancelToken::local().wait(upperAddress(&address), upperValue(value));
+        value = address.load();
+        return res;
+    }
+
+    // inline void wakeUpperOne(std::atomic<UInt64> & address)
+    // {
+    //     syscall(SYS_futex, upperAddress(&address), FUTEX_WAKE_PRIVATE, 1, nullptr, nullptr, 0);
+    // }
+
+    inline void wakeUpperAll(std::atomic<UInt64> & address)
+    {
+        syscall(SYS_futex, upperAddress(&address), FUTEX_WAKE_PRIVATE, INT_MAX, nullptr, nullptr, 0);
+    }
+}
+
+void CancelToken::Registry::insert(CancelToken * token)
+{
+    std::lock_guard<std::mutex> lock(mutex);
+    threads[token->thread_id] = token;
+}
+
+void CancelToken::Registry::remove(CancelToken * token)
+{
+    std::lock_guard<std::mutex> lock(mutex);
+    threads.erase(token->thread_id);
+}
+
+void CancelToken::Registry::signal(UInt64 tid)
+{
+    std::lock_guard<std::mutex> lock(mutex);
+    if (auto it = threads.find(tid); it != threads.end())
+        it->second->signalImpl();
+}
+
+void CancelToken::Registry::signal(UInt64 tid, int code, const String & message)
+{
+    std::lock_guard<std::mutex> lock(mutex);
+    if (auto it = threads.find(tid); it != threads.end())
+        it->second->signalImpl(code, message);
+}
+
+CancelToken::Registry & CancelToken::Registry::instance()
+{
+    static Registry registry;
+    return registry;
+}
+
+CancelToken::CancelToken()
+    : state(disabled)
+    , thread_id(getThreadId())
+{
+    Registry::instance().insert(this);
+}
+
+CancelToken::~CancelToken()
+{
+    Registry::instance().remove(this);
+}
+
+void CancelToken::signal(UInt64 tid)
+{
+    Registry::instance().signal(tid);
+}
+
+void CancelToken::signal(UInt64 tid, int code, const String & message)
+{
+    Registry::instance().signal(tid, code, message);
+}
+
+bool CancelToken::wait(UInt32 * address, UInt32 value)
+{
+    chassert((reinterpret_cast<UInt64>(address) & canceled) == 0); // An `address` must be 2-byte aligned
+    if (value & signaled) // Can happen after spurious wake-up due to cancel of other thread
+        return true; // Spin-wait unless signal is handled
+
+    UInt64 s = state.load();
```

Comment:
`memory_order_acquire`? Why would we need `memory_order_seq_cst` here? I think it'd be better to specify memory ordering explicitly in synchronization primitive implementnation

Reply:
You are right, but seq_cst is the strictest memory order, so it will work fine. Let's call it "room for later optimizations" :)